File: src/app/watch-list.ts
import { jsonArrayFrom } from 'kysely/helpers/postgres';
import { DEFAULT_REPLY_MARKUP } from '../constant';
import { kysely } from '../database';
import type {
  TelegramRequest,
  TelegramResponse,
} from '../types';

export async function watchList(
  req: TelegramRequest,
): Promise<TelegramResponse> {
  const watchList = await kysely
    .selectFrom('watch_list as wl')
    .innerJoin('product', 'product.id', 'wl.product_id')
    .select((eb) => [
      'product.name as product_name',
      jsonArrayFrom(
        eb
          .selectFrom('product_version as pv')
          .select(['pv.version', 'pv.release_date'])
          .whereRef('pv.product_id', '=', 'product.id')
          .orderBy('pv.release_date', 'desc')
          .limit(3),
      ).as('recent_releases'),
    ])
    .where(
      'wl.chat_id',
      '=',
      req.message!.chat.id.toString(),
    )
    .orderBy('product.name')
    .execute();

  let text = '<b>Watch List</b>\n\n';

  if (watchList.length === 0) {
    text += '<i>No products added to watch list</i>';
  } else {
    text += watchList
      .map(
        (p) =>
          `<b>${p.product_name.toUpperCase()}</b>\n` +
          p.recent_releases
            .map(
              (r) =>
                `version: <code>${r.version}</code> - release: ${r.release_date}`,
            )
            .join('\n'),
      )
      .join('\n\n');
  }

  return {
    method: 'sendMessage',
    chat_id: req.message!.chat.id,
    parse_mode: 'HTML',
    text,
    reply_markup: DEFAULT_REPLY_MARKUP,
  };
}

File: internal/handler/watch_list.go
package handler

import (
	"context"
	"fmt"
	"strings"

	"github.com/bytedance/sonic"
	"github.com/fidrasofyan/version-watcher-bot/database"
	"github.com/fidrasofyan/version-watcher-bot/internal/custom_error"
	"github.com/fidrasofyan/version-watcher-bot/internal/service"
	"github.com/fidrasofyan/version-watcher-bot/internal/types"
	"github.com/jackc/pgx/v5/pgtype"
)

type productVersion struct {
	ReleaseLabel       string           `json:"release_label"`
	Version            string           `json:"version"`
	VersionReleaseDate pgtype.Timestamp `json:"version_release_date"`
	VersionReleaseLink *string          `json:"version_release_link"`
}

func WatchList(ctx context.Context, req types.TelegramUpdate) (*types.TelegramResponse, error) {
	watchLists, err := database.Sqlc.GetWatchListsWithProductVersions(ctx, req.Message.Chat.Id)
	if err != nil {
		return nil, custom_error.NewError(err)
	}

	textLimit := 3500
	var textB strings.Builder
	textB.WriteString("<b>Watch List</b>\n")

	switch len(watchLists) {
	case 0:
		textB.WriteString("\n<i>No watch list found</i>")
	case 1:
		textB.WriteString("<i>You watch 1 product</i>\n\n")
	default:
		textB.WriteString(fmt.Sprintf("<i>You watch %d products</i>\n\n", len(watchLists)))
	}

	for _, watchList := range watchLists {
		// Set title
		textB.WriteString(fmt.Sprintf("# <b>%s</b> - <a href=\"%s\">source</a>\n", watchList.ProductLabel, watchList.ProductEolUrl))

		// Set product versions
		productVersions := []productVersion{}
		if len(watchList.ProductVersions) != 0 {
			if err := sonic.Unmarshal(watchList.ProductVersions, &productVersions); err != nil {
				return nil, custom_error.NewError(err)
			}
		}

		for _, pv := range productVersions {
			if pv.VersionReleaseDate.Valid {
				textB.WriteString(fmt.Sprintf("• Latest: %s - %s\n", pv.Version, pv.VersionReleaseDate.Time.Format("2 Jan 2006")))
			} else {
				textB.WriteString("• Latest release: -\n")
			}
		}

		// If text is too long, send it part by part
		if textB.Len() >= textLimit {
			service.SendMessage(&service.SendMessageParams{
				ChatId:             req.Message.Chat.Id,
				ParseMode:          "HTML",
				Text:               textB.String(),
				LinkPreviewOptions: &types.TelegramLinkPreviewOptions{IsDisabled: true},
			})
			textB.Reset()
		}
	}

	if textB.Len() == 0 {
		return nil, nil
	}

	return &types.TelegramResponse{
		Method:      "sendMessage",
		ChatId:      req.Message.Chat.Id,
		ParseMode:   "HTML",
		Text:        textB.String(),
		ReplyMarkup: types.DefaultReplyMarkup,
		LinkPreviewOptions: &types.TelegramLinkPreviewOptions{
			IsDisabled: true,
		},
	}, nil
}
